src/main/java/com/gmail/nossr50/util/blockmeta/chunkmeta/PrimitiveChunkStore.java

```diff
@@ -47,51 +76,75 @@ public int getChunkZ() {
         return cz;
     }
 
+    private boolean isInvalid(int x, int y, int z) {
+        if (y < 0 || y >= worldHeight) return true;
+        return ((x | z) & ~0xF) != 0;
+    }
+
     @Override
     public boolean isTrue(int x, int y, int z) {
-        return store[x][z][y];
+        if (isInvalid(x, y, z)) return false;
+        int status = data.get((y * PAGE_SIZE) + (x >> 1));
+        return ((status >>> (z + (((x & 1) == 1) ? 16 : 0))) & 1) != 0;
```

Comment:
Please include comments inside the method explaining the operations you are taking in this method, especially the bit shifting.